DREAM/Optimization/tsgParticleSwarm.hpp

```diff
@@ -0,0 +1,118 @@
+        particle_positions = pp;
+    }
+    void setParticlePositions(std::vector<double> &&pp) {
+        checkVarSize("ParticleSwarmState::setParticlePositions", "particle positions", pp.size(), num_dimensions * num_particles);
+        particle_positions = std::move(pp);
+    }
+
+    void setParticleVelocities(const std::vector<double> &pv) {
+        checkVarSize("ParticleSwarmState::setParticleVelocities", "particle velocities", pv.size(), num_dimensions * num_particles);
+        particle_velocities = pv;
+    }
+    void setParticleVelocities(std::vector<double> &&pv) {
+        checkVarSize("ParticleSwarmState::setParticleVelocities", "particle velocities", pv.size(), num_dimensions * num_particles);
+        particle_velocities = std::move(pv);
+    }
+
+    void setBestParticlePositions(const std::vector<double> &bpp) {
+        checkVarSize("ParticleSwarmState::setBestParticlePositions", "best particle positions", bpp.size(), num_dimensions * num_particles);
+        best_particle_positions = bpp;
+    }
+    void setBestParticlePositions(std::vector<double> &&bpp) {
+        checkVarSize("ParticleSwarmState::setBestParticlePositions", "best particle positions", bpp.size(), num_dimensions * num_particles);
+        best_particle_positions = std::move(bpp);
+    }
+
+    void setBestPosition(const std::vector<double> &bp) {
+        checkVarSize("ParticleSwarmState::setBestPosition", "best position", bp.size(), num_dimensions);
+        best_position = bp;
+    }
+    void setBestPosition(std::vector<double> &&bp) {
+        checkVarSize("ParticleSwarmState::setBestPosition", "best position", bp.size(), num_dimensions);
+        best_position = std::move(bp);
+    }
+
+    void setNumDimensions(const int nd) {num_dimensions = nd;};
+    void setNumParticles(const int np) {num_particles = np;};
+
+    void setParticlesInsideBox(const std::vector<double> &box_lower, const std::vector<double> &box_upper,
+                               const std::function<double(void)> get_random01 = TasDREAM::tsgCoreUniform01);
+
+    friend void ParticleSwarm(ObjectiveFunction f, int max_iterations, TasDREAM::DreamDomain inside, ParticleSwarmState &state,
+                              double inertia_weight, double cognitive_coeff, double social_coeff, std::function<double(void)> get_random01);
+
+  protected:
+    inline std::vector<double> &getParticlePositionsRef() {return particle_positions;}
+    inline std::vector<double> &getParticleVelocitiesRef() {return particle_velocities;}
+    inline std::vector<double> &getBestParticlePositionsRef() {return best_particle_positions;}
+    inline std::vector<double> &getBestPositionRef() {return best_position;}
+
+  private:
+    bool initialized;
+    int num_dimensions, num_particles;
+    std::vector<double> particle_positions, particle_velocities, best_particle_positions, best_position;
```

Comment:
Values of the objective function should be kept inside the state.

The values are not part of the state, but they are cache data associated with the state.

We should also have a way to reset the values.

Reply:
See comment below.